File: cli/src/commands/reanchor.ts
/**
 * mrsf reanchor — Re-anchor comments after document changes.
 */

import type { Command } from "commander";
import chalk from "chalk";
import path from "node:path";
import { discoverAllSidecars, findWorkspaceRoot } from "../lib/discovery.js";
import { parseSidecar } from "../lib/parser.js";
import { readDocumentLines } from "../lib/parser.js";
import { sidecarToDocument } from "../lib/discovery.js";
import { resolveSidecarPaths } from "../lib/resolve-files.js";
import { writeSidecar } from "../lib/writer.js";
import { findRepoRoot, getStagedFiles } from "../lib/git.js";
import {
  reanchorDocument,
  applyReanchorResults,
} from "../lib/reanchor.js";
import type { ReanchorResult } from "../lib/types.js";

export function registerReanchor(program: Command): void {
  program
    .command("reanchor [files...]")
    .description("Re-anchor comments to current document content")
    .option("-n, --dry-run", "Report without modifying files")
    .option("-t, --threshold <n>", "Fuzzy threshold 0.0–1.0", "0.6")
    .option("--staged", "Only process sidecars for staged files")
    .option("--no-git", "Disable git integration")
    .option("--from <commit>", "Override from-commit for all comments")
    .option("--update-text", "Also update selected_text to match current document text")
    .action(
      async (
        files: string[],
        opts: {
          dryRun?: boolean;
          threshold: string;
          staged?: boolean;
          git: boolean;
          from?: string;
          updateText?: boolean;
        },
      ) => {
        const parentOpts = program.opts();
        const cwd = parentOpts.cwd ?? process.cwd();
        const quiet = parentOpts.quiet ?? false;
        const threshold = parseFloat(opts.threshold);
        const noGit = !opts.git;

        let sidecarPaths: string[];

        if (opts.staged && !noGit) {
          // Only process sidecars whose documents are staged
          const repoRoot = await findRepoRoot(cwd);
          if (!repoRoot) {
            console.error(chalk.red("Not in a git repository."));
            process.exit(1);
          }
          const staged = await getStagedFiles(repoRoot, "*.md");
          sidecarPaths = [];
          for (const md of staged) {
            const absPath = path.resolve(repoRoot, md + ".review.yaml");
            try {
              await parseSidecar(absPath);
              sidecarPaths.push(absPath);
            } catch {
              // No sidecar for this staged file, skip
            }
          }
        } else {
          sidecarPaths = await resolveSidecarPaths(files, cwd);
        }

        if (sidecarPaths.length === 0) {
          if (!quiet) console.log("No sidecar files found.");
          return;
        }

        let totalChanged = 0;
        let totalOrphaned = 0;

        for (const sp of sidecarPaths) {
          try {
            const doc = await parseSidecar(sp);
            const docPath = sidecarToDocument(sp);
            const lines = await readDocumentLines(docPath);
            const repoRoot = !noGit ? await findRepoRoot(cwd) : null;

            const results = await reanchorDocument(doc, lines, {
              threshold,
              noGit,
              fromCommit: opts.from,
              documentPath: docPath,
              repoRoot: repoRoot ?? undefined,
            });

            const orphaned = results.filter((r) => r.status === "orphaned").length;
            totalOrphaned += orphaned;

            if (!opts.dryRun) {
              const changed = applyReanchorResults(doc, results, {
                updateText: opts.updateText,
              });
              totalChanged += changed;
              if (changed > 0) {
                await writeSidecar(sp, doc);
              }
            }

            if (!quiet) {
              printResults(sp, results, opts.dryRun);
            }
          } catch (err: unknown) {
            const msg = err instanceof Error ? err.message : String(err);
            console.error(chalk.red(`✗ ${sp}: ${msg}`));
          }
        }

        if (!quiet) {
          console.log();
          const label = opts.dryRun ? "would change" : "changed";
          console.log(
            `${totalChanged} comment(s) ${label}, ${totalOrphaned} orphaned.`,
          );
        }

        if (totalOrphaned > 0) process.exit(1);
      },
    );
}

function printResults(
  sidecarPath: string,
  results: ReanchorResult[],
  dryRun?: boolean,
): void {
  const prefix = dryRun ? "[dry-run] " : "";
  console.log(`${prefix}${sidecarPath}:`);

  for (const r of results) {
    const icon = statusIcon(r.status);
    const line = r.newLine != null ? `:${r.newLine}` : "";
    console.log(`  ${icon} ${r.commentId}${line} — ${r.reason}`);
  }
}

function statusIcon(status: string): string {
  switch (status) {
    case "anchored":
      return chalk.green("●");
    case "shifted":
      return chalk.blue("↕");
    case "fuzzy":
      return chalk.yellow("≈");
    case "ambiguous":
      return chalk.magenta("?");
    case "orphaned":
      return chalk.red("✗");
    default:
      return "·";
  }
}

File: python/src/mrsf/cli/reanchor.py
"""mrsf reanchor [files...] — re-anchor comments after document edits."""

from __future__ import annotations

import sys

import click

from ..git import find_repo_root, get_staged_files, is_git_available
from ..reanchor import reanchor_file
from ..resolve_files import resolve_sidecar_paths
from ..types import ReanchorOptions
from .main import MrsfContext, pass_ctx


def _status_icon(status: str) -> tuple[str, str]:
    """Return (icon, color) matching the Node.js CLI icons."""
    if status == "anchored":
        return "●", "green"
    if status == "shifted":
        return "↕", "blue"
    if status == "fuzzy":
        return "≈", "yellow"
    if status == "ambiguous":
        return "?", "magenta"
    if status == "orphaned":
        return "✗", "red"
    return "·", "white"


@click.command("reanchor")
@click.argument("files", nargs=-1, type=click.Path())
@click.option("-n", "--dry-run", is_flag=True, help="Report without modifying files.")
@click.option("-t", "--threshold", type=float, default=0.6, help="Fuzzy match threshold (0.0–1.0).")
@click.option("--staged", is_flag=True, help="Reanchor sidecars for staged markdown files.")
@click.option("--no-git", is_flag=True, help="Disable git integration.")
@click.option("--from", "from_commit", default=None, help="Base commit for diff.")
@click.option(
    "--update-text", is_flag=True, help="Update selected_text with current document text."
)
@click.option("-f", "--force", is_flag=True, help="Force-anchor high-confidence results.")
@pass_ctx
def reanchor_cmd(
    ctx: MrsfContext,
    files: tuple[str, ...],
    dry_run: bool,
    threshold: float,
    staged: bool,
    no_git: bool,
    from_commit: str | None,
    update_text: bool,
    force: bool,
) -> None:
    """Re-anchor comments after document edits."""
    if staged:
        if not is_git_available():
            click.echo(click.style("Git is not available.", fg="red"), err=True)
            sys.exit(1)
        repo_root = find_repo_root(ctx.cwd)
        if not repo_root:
            click.echo(click.style("Not in a git repository.", fg="red"), err=True)
            sys.exit(1)
        md_files = [f for f in get_staged_files(repo_root) if f.endswith(".md")]
        if not md_files:
            if not ctx.quiet:
                click.echo("No staged markdown files found.")
            return
        sidecars = resolve_sidecar_paths(md_files, cwd=ctx.cwd)
    else:
        sidecars = resolve_sidecar_paths(list(files), cwd=ctx.cwd)

    if not sidecars:
        click.echo("No sidecar files found.", err=True)
        sys.exit(1)

    opts = ReanchorOptions(
        cwd=ctx.cwd,
        config_path=ctx.config,
        dry_run=dry_run,
        threshold=threshold,
        no_git=no_git,
        from_commit=from_commit,
        update_text=update_text,
        force=force,
    )

    has_orphaned = False
    for sc in sidecars:
        try:
            results, changed, written = reanchor_file(sc, opts)
        except Exception as e:
            click.echo(click.style(f"Error reanchoring {sc}: {e}", fg="red"), err=True)
            continue

        orphaned = sum(1 for r in results if r.status == "orphaned")
        if orphaned:
            has_orphaned = True

        if not ctx.quiet:
            for r in results:
                icon, color = _status_icon(r.status)
                line_info = f" → line {r.new_line}" if r.new_line is not None else ""
                click.echo(
                    click.style(f"  {icon} {r.comment_id}{line_info} ({r.reason})", fg=color)
                )

            action = "would change" if dry_run else "changed"
            click.echo(f"  {sc}: {changed} comment(s) {action}, {orphaned} orphaned")

    if has_orphaned:
        sys.exit(1)
